// ngx-scrollbar/esm2015/lib/classes/scrollbar-ref.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { coerceBooleanProperty, coerceNumberProperty } from '@angular/cdk/coercion';
import { animationFrameScheduler, asyncScheduler, EMPTY, fromEvent, merge, of, Subject } from 'rxjs';
import { distinctUntilChanged, map, mergeMap, pluck, switchMap, takeUntil, tap } from 'rxjs/operators';
/**
 * @abstract
 */
export class ScrollbarRef {
    /**
     * @protected
     * @param {?} scrollbarRef
     * @param {?} document
     * @param {?} trackRef
     * @param {?} thumbRef
     * @param {?} platform
     * @param {?} destroyed
     */
    constructor(scrollbarRef, document, trackRef, thumbRef, platform, destroyed) {
        this.scrollbarRef = scrollbarRef;
        this.document = document;
        this.destroyed = destroyed;
        this.viewElement = scrollbarRef.viewport;
        this.trackElement = trackRef.nativeElement;
        this.thumbElement = thumbRef.nativeElement;
        if (!(platform.IOS || platform.ANDROID)) {
            this.pointerEvents = new Subject();
            this.hoveredState = new Subject();
            this.draggingState = new Subject();
            this.draggingState.pipe(distinctUntilChanged(), tap((/**
             * @param {?} state
             * @return {?}
             */
            (state) => this.setDragging(state))), takeUntil(this.destroyed)).subscribe();
            /** @type {?} */
            const scrollbarClicked = fromEvent(this.viewElement, 'mousedown', { passive: true }).pipe(switchMap((/**
             * @param {?} e
             * @return {?}
             */
            (e) => {
                e.stopPropagation();
                this.document.onselectstart = (/**
                 * @return {?}
                 */
                () => false);
                /** @type {?} */
                const isThumbClick = isWithinBounds(e, this.thumbElement.getBoundingClientRect());
                if (isThumbClick && !coerceBooleanProperty(this.scrollbarRef.thumbDragDisabled)) {
                    return this.dragged(e);
                }
                else {
                    /** @type {?} */
                    const isTrackClick = isWithinBounds(e, this.trackElement.getBoundingClientRect());
                    if (isTrackClick && !coerceBooleanProperty(this.scrollbarRef.trackClickDisabled)) {
                        return this.trackClicked(e);
                    }
                }
                return EMPTY;
            })));
            // Activate/Deactivate scrollbar hover event
            /** @type {?} */
            const mouseLeave = fromEvent(this.viewElement, 'mouseleave').pipe(map((/**
             * @param {?} e
             * @return {?}
             */
            (e) => {
                e.stopPropagation();
                return false;
            })));
            merge(this.pointerEvents, mouseLeave).pipe(distinctUntilChanged()).pipe(tap((/**
             * @param {?} state
             * @return {?}
             */
            (state) => this.setHovered(state))), takeUntil(this.destroyed)).subscribe();
            // Activate/Deactivate scrollTo on scrollbar click event
            this.pointerEvents.pipe(distinctUntilChanged(), switchMap((/**
             * @param {?} state
             * @return {?}
             */
            (state) => state ? scrollbarClicked : EMPTY)), takeUntil(this.destroyed)).subscribe();
            this.hovered().pipe(tap((/**
             * @param {?} state
             * @return {?}
             */
            (state) => this.pointerEvents.next(state))), takeUntil(this.destroyed)).subscribe();
        }
        // Start updating thumb position when view scrolls
        this.scrolled().pipe(tap((/**
         * @return {?}
         */
        () => this.updateThumb())), takeUntil(this.destroyed)).subscribe();
        // Update scrollbar when `NgScrollbar.update()` is called
        this.scrollbarRef.updated.pipe(tap((/**
         * @return {?}
         */
        () => this.updateThumb())), takeUntil(this.destroyed)).subscribe();
        // Initialize scrollbar
        asyncScheduler.schedule((/**
         * @return {?}
         */
        () => this.updateThumb()), 100);
    }
    // The available scrollable size
    /**
     * @protected
     * @return {?}
     */
    get scrollMax() {
        return this.scrollSize - this.viewportSize;
    }
    /**
     * @protected
     * @return {?}
     */
    get trackMax() {
        return this.trackSize - this.thumbSize;
    }
    /**
     * Updates scrollbar's thumb position and size
     * @protected
     * @return {?}
     */
    updateThumb() {
        /** @type {?} */
        const trackMax = this.trackMax;
        /** @type {?} */
        const size = calculateThumbSize(this.trackSize, this.scrollSize, this.scrollbarRef.minThumbSize);
        /** @type {?} */
        const position = calculateThumbPosition(this.scrollOffset, this.scrollMax, trackMax);
        animationFrameScheduler.schedule((/**
         * @return {?}
         */
        () => this.applyThumbStyle(size, position, trackMax)));
    }
    /**
     * @param {?} event
     * @return {?}
     */
    dragged(event) {
        /** @type {?} */
        let trackMax;
        /** @type {?} */
        let scrollMax;
        /** @type {?} */
        const dragStart = of(event).pipe(tap((/**
         * @return {?}
         */
        () => {
            // Capture scrollMax and trackMax once
            trackMax = this.trackMax;
            scrollMax = this.scrollMax;
            this.draggingState.next(true);
        })));
        /** @type {?} */
        const dragging = fromEvent(this.document, 'mousemove', { capture: true, passive: true }).pipe(tap((/**
         * @param {?} e
         * @return {?}
         */
        (e) => e.stopPropagation())));
        /** @type {?} */
        const dragEnd = fromEvent(this.document, 'mouseup', { capture: true }).pipe(tap((/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            e.stopPropagation();
            this.document.onselectstart = null;
            this.draggingState.next(false);
        })));
        return dragStart.pipe(pluck(this.pageProperty), map((/**
         * @param {?} pageOffset
         * @return {?}
         */
        (pageOffset) => pageOffset - this.dragStartOffset)), mergeMap((/**
         * @param {?} mouseDownOffset
         * @return {?}
         */
        (mouseDownOffset) => dragging.pipe(pluck(this.clientProperty), 
        // Calculate how far the user's mouse is from the top/left of the scrollbar (minus the dragOffset).
        map((/**
         * @param {?} mouseOffset
         * @return {?}
         */
        (mouseOffset) => mouseOffset - this.dragOffset)), map((/**
         * @param {?} offset
         * @return {?}
         */
        (offset) => scrollMax * (offset - mouseDownOffset) / trackMax)), map((/**
         * @param {?} position
         * @return {?}
         */
        (position) => this.handleDragPosition(position, scrollMax))), tap((/**
         * @param {?} value
         * @return {?}
         */
        (value) => this.scrollTo(value))), takeUntil(dragEnd)))));
    }
    /**
     * Stream that emits when a scrollbar is hovered
     * @private
     * @return {?}
     */
    hovered() {
        return fromEvent(this.viewElement, 'mousemove', { passive: true }).pipe(map((/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            e.stopPropagation();
            return isWithinBounds(e, this.trackElement.getBoundingClientRect());
        })));
    }
    /**
     * Stream that emits when scrollbar track is clicked
     * @protected
     * @param {?} e
     * @return {?}
     */
    trackClicked(e) {
        return of(e).pipe(pluck(this.pageProperty), map((/**
         * @param {?} pageOffset
         * @return {?}
         */
        (pageOffset) => pageOffset - this.dragOffset)), map((/**
         * @param {?} clickOffset
         * @return {?}
         */
        (clickOffset) => {
            /** @type {?} */
            const offset = clickOffset - (this.thumbSize / 2);
            /** @type {?} */
            const ratio = offset / this.trackSize;
            return ratio * this.scrollSize;
        })), tap((/**
         * @param {?} value
         * @return {?}
         */
        (value) => this.scrollbarRef.scrollTo(Object.assign({}, this.mapToScrollToOption(value), { duration: coerceNumberProperty(this.scrollbarRef.trackClickScrollDuration) })))), tap((/**
         * @return {?}
         */
        () => this.document.onselectstart = null)));
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    ScrollbarRef.prototype.viewElement;
    /**
     * @type {?}
     * @protected
     */
    ScrollbarRef.prototype.trackElement;
    /**
     * @type {?}
     * @protected
     */
    ScrollbarRef.prototype.thumbElement;
    /**
     * @type {?}
     * @protected
     */
    ScrollbarRef.prototype.hoveredState;
    /**
     * @type {?}
     * @protected
     */
    ScrollbarRef.prototype.draggingState;
    /**
     * @type {?}
     * @protected
     */
    ScrollbarRef.prototype.pointerEvents;
    /**
     * @type {?}
     * @protected
     */
    ScrollbarRef.prototype.scrollbarRef;
    /**
     * @type {?}
     * @protected
     */
    ScrollbarRef.prototype.document;
    /**
     * @type {?}
     * @protected
     */
    ScrollbarRef.prototype.destroyed;
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    ScrollbarRef.prototype.scrollSize = function () { };
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    ScrollbarRef.prototype.viewportSize = function () { };
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    ScrollbarRef.prototype.trackSize = function () { };
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    ScrollbarRef.prototype.thumbSize = function () { };
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    ScrollbarRef.prototype.scrollOffset = function () { };
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    ScrollbarRef.prototype.dragStartOffset = function () { };
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    ScrollbarRef.prototype.dragOffset = function () { };
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    ScrollbarRef.prototype.pageProperty = function () { };
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    ScrollbarRef.prototype.clientProperty = function () { };
    /**
     * Stream that emits when view is scrolled
     * @abstract
     * @protected
     * @return {?}
     */
    ScrollbarRef.prototype.scrolled = function () { };
    /**
     * Return a scrollTo option parameter
     * @abstract
     * @protected
     * @param {?} value
     * @return {?}
     */
    ScrollbarRef.prototype.mapToScrollToOption = function (value) { };
    /**
     * Updates scrollbar's thumb size and position
     * @abstract
     * @protected
     * @param {?} size
     * @param {?} position
     * @param {?=} trackMax
     * @return {?}
     */
    ScrollbarRef.prototype.applyThumbStyle = function (size, position, trackMax) { };
    /**
     * On drag function
     * @abstract
     * @protected
     * @param {?} position
     * @param {?} scrollMax
     * @return {?}
     */
    ScrollbarRef.prototype.handleDragPosition = function (position, scrollMax) { };
    /**
     * @abstract
     * @protected
     * @param {?} point
     * @return {?}
     */
    ScrollbarRef.prototype.scrollTo = function (point) { };
    /**
     * @abstract
     * @protected
     * @param {?} value
     * @return {?}
     */
    ScrollbarRef.prototype.setDragging = function (value) { };
    /**
     * @abstract
     * @protected
     * @param {?} value
     * @return {?}
     */
    ScrollbarRef.prototype.setHovered = function (value) { };
}
/**
 * Calculate Scrollbar thumb size
 * @param {?} trackSize Scrollbar track size
 * @param {?} contentSize Content size or Viewport scroll size
 * @param {?} minThumbSize Minimum scrollbar thumb size
 * @return {?}
 */
function calculateThumbSize(trackSize, contentSize, minThumbSize) {
    /** @type {?} */
    const scrollbarRatio = trackSize / contentSize;
    /** @type {?} */
    const thumbSize = scrollbarRatio * trackSize;
    return Math.max(~~thumbSize, minThumbSize);
}
/**
 * Calculate scrollbar thumb position
 * @param {?} scrollPosition The scroll position of the viewport
 * @param {?} scrollMax The max size available to scroll the viewport
 * @param {?} trackMax The max size available to move scrollbar thumb
 * @return {?}
 */
function calculateThumbPosition(scrollPosition, scrollMax, trackMax) {
    return scrollPosition * trackMax / scrollMax;
}
/**
 * Check if pointer is within scrollbar bounds
 * @param {?} e Pointer event
 * @param {?} rect Scrollbar Client Rect
 * @return {?}
 */
function isWithinBounds(e, rect) {
    return (e.clientX >= rect.left &&
        e.clientX <= rect.left + rect.width &&
        e.clientY >= rect.top &&
        e.clientY <= rect.top + rect.height);
}
