// ngx-scrollbar/esm2015/lib/classes/vertical-scrollbar-ref.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ScrollbarRef } from './scrollbar-ref';
export class VerticalScrollbarRef extends ScrollbarRef {
    /**
     * @param {?} scrollbarRef
     * @param {?} document
     * @param {?} trackRef
     * @param {?} thumbRef
     * @param {?} platform
     * @param {?} destroyed
     */
    constructor(scrollbarRef, document, trackRef, thumbRef, platform, destroyed) {
        super(scrollbarRef, document, trackRef, thumbRef, platform, destroyed);
        this.scrollbarRef = scrollbarRef;
        this.document = document;
        this.destroyed = destroyed;
    }
    /**
     * @protected
     * @return {?}
     */
    get scrollSize() {
        return this.viewElement.scrollHeight;
    }
    /**
     * @protected
     * @return {?}
     */
    get viewportSize() {
        return this.viewElement.clientHeight;
    }
    /**
     * @protected
     * @return {?}
     */
    get trackSize() {
        return this.trackElement.clientHeight;
    }
    /**
     * @protected
     * @return {?}
     */
    get thumbSize() {
        return this.thumbElement.clientHeight;
    }
    /**
     * @protected
     * @return {?}
     */
    get scrollOffset() {
        return this.viewElement.scrollTop;
    }
    /**
     * @protected
     * @return {?}
     */
    get dragOffset() {
        return this.trackElement.getBoundingClientRect().top;
    }
    /**
     * @protected
     * @return {?}
     */
    get dragStartOffset() {
        return this.thumbElement.getBoundingClientRect().top;
    }
    /**
     * @protected
     * @return {?}
     */
    get pageProperty() {
        return 'pageY';
    }
    /**
     * @protected
     * @return {?}
     */
    get clientProperty() {
        return 'clientY';
    }
    /**
     * @protected
     * @return {?}
     */
    scrolled() {
        return this.scrollbarRef.verticalScrolled;
    }
    /**
     * @protected
     * @param {?} size
     * @param {?} position
     * @return {?}
     */
    applyThumbStyle(size, position) {
        this.thumbElement.style.height = `${size}px`;
        this.thumbElement.style.transform = `translate3d(0 ,${position}px, 0)`;
    }
    /**
     * @protected
     * @param {?} value
     * @return {?}
     */
    mapToScrollToOption(value) {
        return { top: value };
    }
    /**
     * @protected
     * @param {?} position
     * @return {?}
     */
    handleDragPosition(position) {
        return position;
    }
    /**
     * @protected
     * @param {?} point
     * @return {?}
     */
    scrollTo(point) {
        this.viewElement.scrollTop = point;
    }
    /**
     * @protected
     * @param {?} value
     * @return {?}
     */
    setHovered(value) {
        this.scrollbarRef.setHovered({ verticalHovered: value });
    }
    /**
     * @protected
     * @param {?} value
     * @return {?}
     */
    setDragging(value) {
        this.scrollbarRef.setDragging({ verticalDragging: value });
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    VerticalScrollbarRef.prototype.scrollbarRef;
    /**
     * @type {?}
     * @protected
     */
    VerticalScrollbarRef.prototype.document;
    /**
     * @type {?}
     * @protected
     */
    VerticalScrollbarRef.prototype.destroyed;
}
