// ngx-scrollbar/esm5/lib/classes/vertical-scrollbar-ref.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ScrollbarRef } from './scrollbar-ref';
var VerticalScrollbarRef = /** @class */ (function (_super) {
    tslib_1.__extends(VerticalScrollbarRef, _super);
    function VerticalScrollbarRef(scrollbarRef, document, trackRef, thumbRef, platform, destroyed) {
        var _this = _super.call(this, scrollbarRef, document, trackRef, thumbRef, platform, destroyed) || this;
        _this.scrollbarRef = scrollbarRef;
        _this.document = document;
        _this.destroyed = destroyed;
        return _this;
    }
    Object.defineProperty(VerticalScrollbarRef.prototype, "scrollSize", {
        get: /**
         * @protected
         * @return {?}
         */
        function () {
            return this.viewElement.scrollHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VerticalScrollbarRef.prototype, "viewportSize", {
        get: /**
         * @protected
         * @return {?}
         */
        function () {
            return this.viewElement.clientHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VerticalScrollbarRef.prototype, "trackSize", {
        get: /**
         * @protected
         * @return {?}
         */
        function () {
            return this.trackElement.clientHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VerticalScrollbarRef.prototype, "thumbSize", {
        get: /**
         * @protected
         * @return {?}
         */
        function () {
            return this.thumbElement.clientHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VerticalScrollbarRef.prototype, "scrollOffset", {
        get: /**
         * @protected
         * @return {?}
         */
        function () {
            return this.viewElement.scrollTop;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VerticalScrollbarRef.prototype, "dragOffset", {
        get: /**
         * @protected
         * @return {?}
         */
        function () {
            return this.trackElement.getBoundingClientRect().top;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VerticalScrollbarRef.prototype, "dragStartOffset", {
        get: /**
         * @protected
         * @return {?}
         */
        function () {
            return this.thumbElement.getBoundingClientRect().top;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VerticalScrollbarRef.prototype, "pageProperty", {
        get: /**
         * @protected
         * @return {?}
         */
        function () {
            return 'pageY';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VerticalScrollbarRef.prototype, "clientProperty", {
        get: /**
         * @protected
         * @return {?}
         */
        function () {
            return 'clientY';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @protected
     * @return {?}
     */
    VerticalScrollbarRef.prototype.scrolled = /**
     * @protected
     * @return {?}
     */
    function () {
        return this.scrollbarRef.verticalScrolled;
    };
    /**
     * @protected
     * @param {?} size
     * @param {?} position
     * @return {?}
     */
    VerticalScrollbarRef.prototype.applyThumbStyle = /**
     * @protected
     * @param {?} size
     * @param {?} position
     * @return {?}
     */
    function (size, position) {
        this.thumbElement.style.height = size + "px";
        this.thumbElement.style.transform = "translate3d(0 ," + position + "px, 0)";
    };
    /**
     * @protected
     * @param {?} value
     * @return {?}
     */
    VerticalScrollbarRef.prototype.mapToScrollToOption = /**
     * @protected
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return { top: value };
    };
    /**
     * @protected
     * @param {?} position
     * @return {?}
     */
    VerticalScrollbarRef.prototype.handleDragPosition = /**
     * @protected
     * @param {?} position
     * @return {?}
     */
    function (position) {
        return position;
    };
    /**
     * @protected
     * @param {?} point
     * @return {?}
     */
    VerticalScrollbarRef.prototype.scrollTo = /**
     * @protected
     * @param {?} point
     * @return {?}
     */
    function (point) {
        this.viewElement.scrollTop = point;
    };
    /**
     * @protected
     * @param {?} value
     * @return {?}
     */
    VerticalScrollbarRef.prototype.setHovered = /**
     * @protected
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.scrollbarRef.setHovered({ verticalHovered: value });
    };
    /**
     * @protected
     * @param {?} value
     * @return {?}
     */
    VerticalScrollbarRef.prototype.setDragging = /**
     * @protected
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.scrollbarRef.setDragging({ verticalDragging: value });
    };
    return VerticalScrollbarRef;
}(ScrollbarRef));
export { VerticalScrollbarRef };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    VerticalScrollbarRef.prototype.scrollbarRef;
    /**
     * @type {?}
     * @protected
     */
    VerticalScrollbarRef.prototype.document;
    /**
     * @type {?}
     * @protected
     */
    VerticalScrollbarRef.prototype.destroyed;
}
