// ngx-scrollbar/esm5/lib/utils/native-scrollbar-size-factory.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { Platform } from '@angular/cdk/platform';
import { BehaviorSubject, fromEvent, of } from 'rxjs';
import { debounceTime, tap, switchMap } from 'rxjs/operators';
import { ScrollbarManager } from './scrollbar-manager';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "./scrollbar-manager";
import * as i3 from "@angular/cdk/platform";
var NativeScrollbarSizeFactory = /** @class */ (function () {
    function NativeScrollbarSizeFactory(document, manager, platform) {
        var _this = this;
        this.document = document;
        this.manager = manager;
        this.platform = platform;
        if (platform.isBrowser) {
            of(null).pipe(tap((/**
             * @return {?}
             */
            function () { return _this._nativeScrollbarSize = new BehaviorSubject(_this.getNativeScrollbarSize()); })), tap((/**
             * @return {?}
             */
            function () { return _this.nativeScrollbarSize = _this._nativeScrollbarSize.asObservable(); })), switchMap((/**
             * @return {?}
             */
            function () { return fromEvent(_this.document.defaultView, 'resize', { passive: true }); })), debounceTime(this.manager.globalOptions.windowResizeDebounce), tap((/**
             * @return {?}
             */
            function () { return _this._nativeScrollbarSize.next(_this.getNativeScrollbarSize()); }))).subscribe();
        }
    }
    /**
     * Get native scrollbar size
     */
    /**
     * Get native scrollbar size
     * @private
     * @return {?}
     */
    NativeScrollbarSizeFactory.prototype.getNativeScrollbarSize = /**
     * Get native scrollbar size
     * @private
     * @return {?}
     */
    function () {
        if (this.platform.ANDROID || this.platform.IOS) {
            return 6;
        }
        /** @type {?} */
        var box = this.document.createElement('div');
        box.className = 'ng-scrollbar-measure';
        this.document.body.appendChild(box);
        /** @type {?} */
        var size = box.getBoundingClientRect().right;
        this.document.body.removeChild(box);
        // if size is 0, return 15 (for MAC OS browsers)
        return size || 15;
    };
    NativeScrollbarSizeFactory.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    NativeScrollbarSizeFactory.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: ScrollbarManager },
        { type: Platform }
    ]; };
    /** @nocollapse */ NativeScrollbarSizeFactory.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function NativeScrollbarSizeFactory_Factory() { return new NativeScrollbarSizeFactory(i0.ɵɵinject(i1.DOCUMENT), i0.ɵɵinject(i2.ScrollbarManager), i0.ɵɵinject(i3.Platform)); }, token: NativeScrollbarSizeFactory, providedIn: "root" });
    return NativeScrollbarSizeFactory;
}());
export { NativeScrollbarSizeFactory };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NativeScrollbarSizeFactory.prototype._nativeScrollbarSize;
    /** @type {?} */
    NativeScrollbarSizeFactory.prototype.nativeScrollbarSize;
    /**
     * @type {?}
     * @private
     */
    NativeScrollbarSizeFactory.prototype.document;
    /**
     * @type {?}
     * @private
     */
    NativeScrollbarSizeFactory.prototype.manager;
    /**
     * @type {?}
     * @private
     */
    NativeScrollbarSizeFactory.prototype.platform;
}
