// ngx-scrollbar/esm5/lib/scrollbar-control/scrollbar-control.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Input, Inject, NgZone, ElementRef, ChangeDetectionStrategy } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { Platform } from '@angular/cdk/platform';
import { Directionality } from '@angular/cdk/bidi';
import { Subject } from 'rxjs';
import { NgScrollbar } from '../ng-scrollbar';
import { HorizontalScrollbarRef } from '../classes/horizontal-scrollbar-ref';
import { VerticalScrollbarRef } from '../classes/vertical-scrollbar-ref';
var ScrollbarControl = /** @class */ (function () {
    function ScrollbarControl(parent, platform, dir, zone, document) {
        this.parent = parent;
        this.platform = platform;
        this.dir = dir;
        this.zone = zone;
        this.document = document;
        // Stream that emits to unsubscribe from all streams
        this.destroyed = new Subject();
    }
    /**
     * @return {?}
     */
    ScrollbarControl.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.zone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            // Avoid SSR Error
            if (_this.platform.isBrowser) {
                _this.scrollbarRef = _this.track === 'vertical'
                    ? new VerticalScrollbarRef(_this.parent, _this.document, _this.trackRef, _this.thumbRef, _this.platform, _this.destroyed)
                    : new HorizontalScrollbarRef(_this.parent, _this.document, _this.trackRef, _this.thumbRef, _this.platform, _this.destroyed, _this.dir);
            }
        }));
    };
    /**
     * @return {?}
     */
    ScrollbarControl.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.destroyed.next();
        this.destroyed.complete();
    };
    ScrollbarControl.decorators = [
        { type: Component, args: [{
                    selector: 'scrollbar-control',
                    host: {
                        '[class.scrollbar-control]': 'true'
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: "\n    <div #track class=\"ng-scrollbar-track {{parent.trackClass}}\">\n      <div #thumb class=\"ng-scrollbar-thumb {{parent.thumbClass}}\"></div>\n    </div>\n  ",
                    styles: ["::ng-deep .ng-scrollbar-thumb,::ng-deep .ng-scrollbar-track,::ng-deep .scrollbar-control{pointer-events:none}::ng-deep .ng-scrollbar-wrapper>.scrollbar-control{position:absolute;display:flex;justify-content:center;align-items:center;transition:height 150ms ease-out,width 150ms ease-out}::ng-deep .ng-scrollbar-wrapper>.scrollbar-control[scrollable=false] .ng-scrollbar-thumb{display:none}::ng-deep .ng-scrollbar-track{height:100%;width:100%;z-index:1;border-radius:var(--scrollbar-border-radius);background-color:var(--scrollbar-track-color);overflow:hidden;transition:height 150ms ease-out,width 150ms ease-out;cursor:default}::ng-deep .ng-scrollbar-thumb{box-sizing:border-box;position:relative;border-radius:inherit;background-color:var(--scrollbar-thumb-color);transform:translateZ(0);transition:height 150ms ease-out,width 150ms ease-out}", "::ng-deep .ng-scrollbar-wrapper>.scrollbar-control[track=horizontal]{height:var(--horizontal-scrollbar-total-size)}::ng-deep .ng-scrollbar-wrapper>.scrollbar-control[track=horizontal]>.ng-scrollbar-track{height:var(--horizontal-scrollbar-size);width:calc(100% - var(--scrollbar-padding) * 2)}::ng-deep .ng-scrollbar-wrapper>.scrollbar-control[track=horizontal]>.ng-scrollbar-track>.ng-scrollbar-thumb{width:0;height:100%}::ng-deep .ng-scrollbar-wrapper[horizontalDragging=true]>.scrollbar-control[track=horizontal] .ng-scrollbar-thumb,::ng-deep .ng-scrollbar-wrapper[horizontalHovered=true]>.scrollbar-control[track=horizontal] .ng-scrollbar-thumb{background-color:var(--scrollbar-thumb-hover-color)}::ng-deep .ng-scrollbar-wrapper[position=invertAll]>.scrollbar-control[track=horizontal],::ng-deep .ng-scrollbar-wrapper[position=invertX]>.scrollbar-control[track=horizontal]{top:0;bottom:unset}::ng-deep .ng-scrollbar-wrapper[disabled=false]>.scrollbar-control[track=horizontal]{left:0;right:0;bottom:0;top:unset}::ng-deep .ng-scrollbar-wrapper[disabled=false][position=invertAll]>.scrollbar-control[track=horizontal],::ng-deep .ng-scrollbar-wrapper[disabled=false][position=invertX]>.scrollbar-control[track=horizontal]{top:0;bottom:unset}::ng-deep .ng-scrollbar-wrapper[disabled=false][track=all][dir=ltr]>.scrollbar-control[track=horizontal][fit=true]{right:var(--scrollbar-total-size);left:0}::ng-deep .ng-scrollbar-wrapper[disabled=false][track=all][dir=ltr][position=invertAll]>.scrollbar-control[track=horizontal][fit=true],::ng-deep .ng-scrollbar-wrapper[disabled=false][track=all][dir=ltr][position=invertY]>.scrollbar-control[track=horizontal][fit=true]{left:var(--scrollbar-total-size);right:0}::ng-deep .ng-scrollbar-wrapper[disabled=false][track=all][dir=rtl]>.scrollbar-control[track=horizontal][fit=true]{left:var(--scrollbar-total-size);right:0}::ng-deep .ng-scrollbar-wrapper[disabled=false][track=all][dir=rtl][position=invertAll]>.scrollbar-control[track=horizontal][fit=true],::ng-deep .ng-scrollbar-wrapper[disabled=false][track=all][dir=rtl][position=invertY]>.scrollbar-control[track=horizontal][fit=true]{right:var(--scrollbar-total-size);left:0}", "::ng-deep .ng-scrollbar-wrapper>.scrollbar-control[track=vertical]{width:var(--vertical-scrollbar-total-size)}::ng-deep .ng-scrollbar-wrapper>.scrollbar-control[track=vertical]>.ng-scrollbar-track{width:var(--vertical-scrollbar-size);height:calc(100% - var(--scrollbar-padding) * 2)}::ng-deep .ng-scrollbar-wrapper>.scrollbar-control[track=vertical]>.ng-scrollbar-track>.ng-scrollbar-thumb{height:0;width:100%}::ng-deep .ng-scrollbar-wrapper[verticalDragging=true]>.scrollbar-control[track=vertical] .ng-scrollbar-thumb,::ng-deep .ng-scrollbar-wrapper[verticalHovered=true]>.scrollbar-control[track=vertical] .ng-scrollbar-thumb{background-color:var(--scrollbar-thumb-hover-color)}::ng-deep .ng-scrollbar-wrapper[disabled=false]>.scrollbar-control[track=vertical]{top:0;bottom:0}::ng-deep .ng-scrollbar-wrapper[disabled=false][dir=ltr]>.scrollbar-control[track=vertical]{right:0;left:unset}::ng-deep .ng-scrollbar-wrapper[disabled=false][dir=ltr][position=invertAll]>.scrollbar-control[track=vertical],::ng-deep .ng-scrollbar-wrapper[disabled=false][dir=ltr][position=invertY]>.scrollbar-control[track=vertical]{left:0;right:unset}::ng-deep .ng-scrollbar-wrapper[disabled=false][dir=rtl]>.scrollbar-control[track=vertical]{left:0;right:unset}::ng-deep .ng-scrollbar-wrapper[disabled=false][dir=rtl][position=invertAll]>.scrollbar-control[track=vertical],::ng-deep .ng-scrollbar-wrapper[disabled=false][dir=rtl][position=invertY]>.scrollbar-control[track=vertical]{left:unset;right:0}::ng-deep .ng-scrollbar-wrapper[disabled=false][track=all]>.scrollbar-control[track=vertical][fit=true]{bottom:var(--scrollbar-total-size);top:0}::ng-deep .ng-scrollbar-wrapper[disabled=false][track=all][position=invertAll]>.scrollbar-control[track=vertical][fit=true],::ng-deep .ng-scrollbar-wrapper[disabled=false][track=all][position=invertX]>.scrollbar-control[track=vertical][fit=true]{top:var(--scrollbar-total-size);bottom:0}"]
                }] }
    ];
    /** @nocollapse */
    ScrollbarControl.ctorParameters = function () { return [
        { type: NgScrollbar },
        { type: Platform },
        { type: Directionality },
        { type: NgZone },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
    ]; };
    ScrollbarControl.propDecorators = {
        track: [{ type: Input }],
        trackRef: [{ type: ViewChild, args: ['track', { static: true },] }],
        thumbRef: [{ type: ViewChild, args: ['thumb', { static: true },] }]
    };
    return ScrollbarControl;
}());
export { ScrollbarControl };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ScrollbarControl.prototype.scrollbarRef;
    /**
     * @type {?}
     * @private
     */
    ScrollbarControl.prototype.track;
    /** @type {?} */
    ScrollbarControl.prototype.trackRef;
    /** @type {?} */
    ScrollbarControl.prototype.thumbRef;
    /**
     * @type {?}
     * @protected
     */
    ScrollbarControl.prototype.destroyed;
    /** @type {?} */
    ScrollbarControl.prototype.parent;
    /**
     * @type {?}
     * @private
     */
    ScrollbarControl.prototype.platform;
    /**
     * @type {?}
     * @private
     */
    ScrollbarControl.prototype.dir;
    /**
     * @type {?}
     * @private
     */
    ScrollbarControl.prototype.zone;
    /**
     * @type {?}
     * @private
     */
    ScrollbarControl.prototype.document;
}
